scripts/build_plugin.py

```diff
@@ -0,0 +1,484 @@
+
+        fh.write("# this file was auto generated.\n")
+
+        fh.write("\nbase_configuration=\"%s\"\n\n" % cfg_descriptor.get_uri())
+
+        for (parameter, value) in manifest_data.iteritems():
+
+            if parameter == "base_configuration":
+                # configuration is processed separately
+                continue
+
+            if isinstance(value, str):
+               fh.write("%s=\"%s\"\n" % (parameter, value.replace("\"", "'")))
+            elif isinstance(value, int):
+                fh.write("%s=%d\n" % (parameter, value))
+            elif isinstance(value, bool):
+                fh.write("%s=%s\n" % (parameter, value))
+            else:
+                raise ValueError(
+                    "Invalid manifest value %s: %s - data type not supported!" % (parameter, value)
+                )
+
+        fh.write("\n\n# system generated parameters\n")
+        fh.write(
+            "BUILD_INFO=\"%s %s@%s\"\n" % (
+                datetime.datetime.now().strftime("%Y%m%d_%H%M%S"),
+                getpass.getuser(),
+                socket.getfqdn(),
+            )
+        )
+        fh.write("BUILD_GENERATION=%d\n" % BUILD_GENERATION)
+
+        # write out helper method add_sgtk_to_pythonpath()
+        core_python_path = os.path.join(core_descriptor.get_path(), "python")
+        core_python_path_relative = core_python_path[len(plugin_root)+1:]
+
+
+        fh.write("def add_sgtk_to_pythonpath(plugin_root):\n")
+        fh.write("    import sys\n")
+        fh.write("    import os\n")
+        fh.write("    core_path = os.path.join(plugin_root, '%s')\n" % core_python_path_relative)
+        fh.write("    sys.path.append(core_path)\n")
+        fh.write("    return core_path\n")
+        fh.write("\n\n")
+
+
+        fh.write("\n# end of file.\n")
+
+def build_plugin(sg_connection, source_path, target_path):
+    """
+    Perform a build of a plugin.
+
+    This will introspect the info.yml in the source path,
+    copy over everything in the source path into target path
+    and then establish a bundle cache containing a reflection
+    of all items required by the config.
+
+    :param sg_connection: Shotgun connection
+    :param source_path: Path to plugin.
+    :param target_path: Path to build
+    """
+    logger.info("Your toolkit plugin in '%s' will be processed." % source_path)
+    logger.info("The build will generated into '%s'" % target_path)
+
+    # check for existence
+    if not os.path.exists(source_path):
+        raise TankError("Source path '%s' cannot be found on disk!" % source_path)
+
+    # check that target path doesn't exist
+    if os.path.exists(target_path):
+        logger.info("The folder '%s' already exists on disk." % target_path)
+        timestamp = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
+        backup_path = "%s.%s" % (target_path, timestamp)
+        filesystem.move_folder(target_path, backup_path)
+        logger.info("...moved it to '%s'" % backup_path)
```

Comment:
minor: I've seen this in a few places in core. Seems like it might be nice to have a utility method to backup a directory using this logic. Maybe a separate ticket? Just a thought.